src/acl_thread.cpp

```diff
@@ -70,6 +68,44 @@ void acl_wait_for_device_update(cl_context context) {
 
 void acl_signal_device_update() { acl_signal_condvar(&l_acl_global_condvar); }
 
+acl_wrapper_mutex::acl_wrapper_mutex() { int id = 0; }
+
+acl_wrapper_mutex::~acl_wrapper_mutex() {}
+
+acl_wrapper_mutex::acl_wrapper_mutex(int idx) : id(idx) {}
+
+void acl_wrapper_mutex::lock() { acl_lock(); }
+
+void acl_wrapper_mutex::unlock() { acl_unlock(); }
+
+// note: acl_threadsupport does not provide this functionality
+// therefore only returning false
+// might use std::lock_gaurd instead
+bool acl_wrapper_mutex::try_lock() { return false; }
+
+acl_condition_lock_guard::acl_condition_lock_guard(acl_wrapper_mutex &mtex,
+                                                   bool cond)
+    : mtex(mtex), cond(cond) {
+  if (cond) {
+    mtex.lock();
+  }
+}
+
+acl_condition_lock_guard::~acl_condition_lock_guard() {
+  if (cond) {
+    mtex.unlock();
+  }
+}
+
+acl_suspend_lock_guard::acl_suspend_lock_guard(acl_wrapper_mutex &mtex)
+    : mtex(mtex) {
+  lock_count = acl_suspend_lock();
+}
+
+acl_suspend_lock_guard::~acl_suspend_lock_guard() {
+  acl_resume_lock(lock_count);
+}
```

Comment:
Sure, I'll put them in the class definition. Would adding the inline option be necessary tho?

Reply:
Okay, I see. Thanks for the detailed explanation Peter!

Reply:
One issue though. Since the class lies in the acl_thread.h header file and calls acl_lock() and acl_unlock(), doesn't this create circular dependency? How should I overcome this? One solution I guess is to create another header file called something like "acl_threadenhanced.hpp" but that would likely be more messy.

Reply:
Oh, sorry. This was a stupid question...coding in early morning didn't exactly work out...